python-checks/src/main/java/org/sonar/python/checks/DeadStoreCheck.java

```diff
@@ -157,60 +166,81 @@ private static boolean isFunctionDeclarationSymbol(Symbol symbol) {
+    Siblings siblings = Siblings.explore(assignmentElement);
 
-  private static Optional<Token> separatorTokenOfElement(Tree element) {
-    if (element.is(Tree.Kind.ASSIGNMENT_STMT, Tree.Kind.EXPRESSION_STMT)) {
-      Token separator = ((Statement) element).separator();
-      return Optional.ofNullable(separator);
+    // Statement is the single element in the block
+    // Replace by `pass` keyword
+    if (siblings.previous == null && siblings.next == null) {
+      return PythonTextEdit.replace(assignmentElement, "pass");
     }
-    return Optional.empty();
-  }
 
-  private static PythonTextEdit removeDeadStore(Tree currentTree) {
-    Optional<Token> tokenSeparator = separatorTokenOfElement(currentTree);
-    List<Tree> childrenOfParent = currentTree.parent().children();
+    boolean hasPreviousElementOnLine = hasPreviousElementOnLine(firstTokenOfElement, siblings.previous);
+    boolean hasNextElementOnLine = hasNextElementOnLine(lastTokenOfElement, siblings.next);
+
+    if (hasNextElementOnLine) {
+      // Statement is first element on the line or between at least two elements
+      // Remove from first token to last toke of element
+      Token firstNextToke = siblings.next.firstToken();
+      return PythonTextEdit.removeRange(firstTokenOfElement.line(), firstTokenOfElement.column(), firstNextToke.line(), firstNextToke.column());
+    } else if (hasPreviousElementOnLine) {
+      // Statement is last element on the line and has one or more previous element on the line
+      // Remove from last token or separator of previous element to avoid trailing white spaces
+      // Keep the line break to leave previous element on same line
+      Token lastPreviousToken = getElementSeparatorOrLastToken(siblings.previous);
+      return PythonTextEdit.removeRange(lastPreviousToken.line(), getEndColumn(lastPreviousToken), lastPreviousToken.line(), getEndColumn(lastTokenOfElement) - 1);
+    } else {
+      // Statement is single element on the line
+      // Remove the entire line including indent and line break
+      return PythonTextEdit.removeRange(firstTokenOfElement.line(), 0, lastTokenOfElement.line(), getEndColumn(lastTokenOfElement));
+    }
+  }
 
-    if (childrenOfParent.size() == 1) {
-      return PythonTextEdit.replace(currentTree, "pass");
+  /**
+   * Statements can have a separator like semicolon. When handling ranges we want to take them into account.
+   */
+  private static Token getElementSeparatorOrLastToken(Tree tree) {
+    if (tree instanceof Statement) {
+      Token separator = ((Statement) tree).separator();
+      if (separator != null) {
+        return separator;
+      }
     }
+    return tree.lastToken();
+  }
 
-    Token currentFirstToken = currentTree.firstToken();
-    int indexOfCurrentTree = childrenOfParent.indexOf(currentTree);
+  private static boolean hasPreviousElementOnLine(Token firstTokenOfElement, @Nullable Tree previousTree) {
```

Comment:
This is purely personal preference, but I have a feeling the `removeStatement` method is a bit easier to understand when I inline the `hasPreviousElementOnLine` and `hasNextElementOnLine` methods, as they are fairly simple. Feel free to disregard, as I understand that others may feel the exact opposite way 😄.